[#51188] Split "% Complete" into a derived-value and self-value

aaron-contreras opened the pull request:
See: https://community.openproject.org/work_packages/51188

This PR will consist of multiple implementation Work Packages and merged in as a single feature. So far, it starts with the following:

- [ ] https://community.openproject.org/projects/openproject/work_packages/50957
- [ ] https://community.openproject.org/projects/openproject/work_packages/51187

aaron-contreras committed:
Generate migration to add `derived_done_ratio` to work packages

* Runs `bin/rails generate migration`.

* Adds column to `work_packages` and `work_package_journals`.

* Backfills column for non leaf work packages.

aaron-contreras committed:
Add `derived_done_ratio` to the work package base contract

I prefer having this explicitly marked as false than having to infer
that because of the attribute not being listed, it's non-writable.

aaron-contreras committed:
Add `derived_done_ratio` to the work package schema representer

Gives the client knowledge of this field in order to pick up on it
and render it within forms.

aaron-contreras committed:
Add `derived_done_ratio` to the work package representer

Returns the derived_done_ratio in the response when fetching a work
package via the API.

aaron-contreras committed:
Add attribute translation for `derived_done_ratio`

Provides the localized name for this field.

aaron-contreras committed:
Add `derivedPercentageDone` to our OAS3 documentation

Keeping docs up-to-date.

aaron-contreras committed:
Complete the range constraints for the OAS specs for `percentageDone`

aaron-contreras committed:
Chore: Alphabetize imports in `display-field.initializer.ts`

aaron-contreras committed:
Render derived_done_ratio alongside done_ratio in cell columns

aaron-contreras committed:
Fix work package pdf export spec

With the added column to the work packages table, a new page was
rendered when exporting the PDF under test. This is a non-functional
change and simply situational due to the newly added column.

Adapted the expected output to match what is being outputted now by
the renderer.

aaron-contreras committed:
Update `derived_done_ratio` on update_ancestores callback

* Instead of modifying the `done_ratio` of other work packages in the
  chain, this commit modifies the `derived_done_ratio` of all relevant
  work packages in the chain is this is now the value that is derived.

* Allows `done_ratio` to be writable as long as the setting allows it.
  There is no longer a restriction on the work package being a leaf
  to be writable.

aaron-contreras committed:
Lint `work_packages/update_service_integration_spec`

* Runs `bundle exec rubocop -a` on spec file and corrects offenses.

cbliard commented on db/migrate/20240116165933_add_derived_done_ratio_to_work_packages.rb:
```diff
@@ -0,0 +1,66 @@
+
+  def migrate_to_derived_done_ratio
+    non_leaf_work_packages
+      .where.not(done_ratio: nil)
```
Not sure if this condition is actually useful: the `done_ratio` field has a non-null constraint and defaults to 0.

cbliard commented on db/migrate/20240116165933_add_derived_done_ratio_to_work_packages.rb:
```diff
@@ -0,0 +1,66 @@
+
+  # Sets the SystemUser as the author of the Journal
+  # entries generated by this migration
+  def journal_user = SystemUser.first
```
Using `User.system` is better, as it will create the system user if it does not exist, although it should be already created at this point.

cbliard commented on app/contracts/work_packages/base_contract.rb:
```diff
@@ -56,6 +56,8 @@ class BaseContract < ::ModelContract
               writable: ->(*) {
                 model.leaf? && Setting.work_package_done_ratio == 'field'
               }
+    attribute :derived_done_ratio,
+              writable: false
```
> I prefer having this explicitly marked as false than having to infer that because of the attribute not being listed, it's non-writable.

I totally agree. That was also my reasoning with `:derived_remaining_hours`.

cbliard replied:
And a test in `lib/api/v3/work_packages/schema/work_package_schema_representer.rb` would be nice, to ensure that making it writable by accident would be catched by a unit test.

cbliard commented on spec/services/work_packages/update_service_integration_spec.rb:
```diff
@@ -413,34 +415,29 @@
-        .to eql(attributes[:start_date])
-      expect(work_package.due_date)
-        .to eql(attributes[:due_date])
+      expect(work_package)
+        .to have_attributes(start_date: attributes[:start_date],
+                            due_date: attributes[:due_date])
```
🖌️ 
I too like the `have_attributes` matcher. Especially when the test fails, the expectation message feels nicer (even if it is sometimes a bit messy when objects have a lot of instance variables).

cbliard reviewed (changes requested):
Nice and clean. Congrats!

I think a few things can be improved:
- the derived % complete should be hidden for leaves work package.
- the derived value should not look like a link, or it should be a proper link, but blue with underline on hover makes it feel like a link and clicking it is deceiving this expectation as nothing happens.

Once done you can merge it.

aaron-contreras committed:
Remove non-null query constraint on derived_done_ratio migrations

* There is no scenario where this column is NULL as the column default
  is 0.

aaron-contreras committed:
Use `User.system` instead of `System.user` for migration journal user

* Ensures the system user is created if it doesn't already exist.

aaron-contreras committed:
Add schema spec for `derivedPercentageDone`

Ensures qualitative properties of the field.

aaron-contreras committed:
Render derived percentage done only if derivable

If a work package is a leaf, there is no done_ratio to derive.

Hence, we'd only like to render the derived_done_ratio if the work
package has children in its hierarchy.

cbliard commented:
Another thing I have just seen while working on another work package: the derived % complete is now available in the form configuration of types.

![image](https://github.com/opf/openproject/assets/176055/075ecd7a-86ac-4165-a276-35df6c84e0cb)

It should not be there.

I now understand that that's how the derived remaining work got added by accident. As soon as I found how to exclude it, I'll share the method to do so with you here.

EDIT: it needs to be added to the `EXCLUDED` array in `app/models/type/attributes.rb`.

aaron-contreras committed:
Add link to derived done ratio query

Navigates to a query with a view of descendant work packages to
inspect the breakdown of child completion.

aaron-contreras committed:
Adapt update_ancestors feature spec expectations

We now expect the derived_done_ratio to be updated instead of the
parent's own done_ratio.

aaron-contreras committed:
Remove `journal_user` definition from migration

Creating activity entries just results in noise and is not being used
at the moment as using update_all bypasses Rails hooks.

aaron-contreras commented:
I've addressed your feedback :) @cbliard

cbliard commented on frontend/src/app/shared/components/fields/display/field-types/compound-progress-display-field.module.ts:
```diff
@@ -0,0 +1,140 @@
+    element.appendChild(span);
+  }
+
+  private addURLToViewWorkPackageChildren(link:HTMLAnchorElement):void {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (this.resource && this.resource.id && this.resource.project) {
+      const wpID = this.resource.id.toString();
+      this
+        .apiV3Service
+        .projects
+        .id(this.resource.project as ProjectResource)
+        .get()
+        .subscribe((project:ProjectResource) => {
+          const props = {
+            c: ['id', 'subject', 'type', 'status', 'estimatedTime', 'remainingTime', 'percentageDone'],
+            hi: true,
+            is: true,
+            f: [{ n: 'parent', o: '=', v: [wpID] }],
+          };
+          const href = URI(this.PathHelper.projectWorkPackagesPath(project.identifier as string))
+            .query({ query_props: JSON.stringify(props) })
+            .toString();
+
+          link.href = href;
+        });
+    }
+  }
```
The links to view work estimates of work package and its children should show the same query regardless of clinking on the derived work, remaining work or % complete.

The simplest to do so would be to have the code for `addURLToViewWorkPackageChildren` in `EstimatedTimeDisplayField` and `CompoundProgressDisplayField` deduplicated.

aaron-contreras replied:
Okay, I interpreted it as clicking on work related links leads to a work-focused query (sans % complete), and clicking on the % complete link leads to a query including the % complete column. Thanks for clarifying

cbliard reviewed (changes requested):
Thanks for addressing my previous remarks 👍 

There are two things remaining: 
- the code to generate the url is the same as for the work/remaining work and should be factored out.
- the attribute `derived_percentage_done` must not appear in the work package form. To exclude it from the form, add the field name to `EXCLUDED` in `app/models/type/attributes.rb`.

aaron-contreras committed:
Extract href generator for derived value query

De-duplicates the generation as the links are identical for both
DisplayField modules.

aaron-contreras committed:
Exclude `derived_percentage_done` from the WorkPackage type form

aaron-contreras committed:
Revert PDF export spec adaptation

Since the derived % complete is no longer included in the type form
configuration, it's no longer included in the pdf export, so the need
for adapting the spec in this PR is no longer relevant.

cbliard committed:
Remove left over comment

cbliard reviewed (approved):
Nice!
